### PR TITLE
sdk/state: move iteration to be pulled from latestCloseAgreement

### PR DESCRIPTION
**Pull request by acharb**
removes `c.SetIterationNumber`. The channel's current iteration number = lastCloseAgreement.IterationNumber + 1

note:
- not updating `lastUnconfirmedPayment` yet. So adding the field as a dummy field for when it's added

closes https://github.com/stellar/experimental-payment-channels/issues/64 and https://github.com/stellar/experimental-payment-channels/issues/61

### CODE REVIEW

**Comment by acharb** [comment]
@leighmcculloch the validation might be oversimplified. Basically I'm saying in `ConfirmPayment` if my channel is expecting a certain iteration number and the payment you gave me is different, then return an error.

That validation might get more complicated when `lastUnconfirmedPayment` comes into play 🤔 . But atm this seems right to me

File: sdk/state/update.go [hunk 1]
```diff
@@ -91,6 +92,9 @@ func (c *Channel) PaymentTxs(p *Payment) (close, decl *txnbuild.Transaction, err
 }
 
 func (c *Channel) ConfirmPayment(p *Payment) (payment *Payment, fullySigned bool, err error) {
+	if p.IterationNumber != c.IterationNumber() {
+		return nil, fullySigned, errors.New("invalid payment iteration number")
```
Comment (leighmcculloch) [inline sdk/state/update.go:96]: Can you include the got and want values in this error message? It'll help with debugging I'm guessing.

**Review by leighmcculloch — APPROVED** [review]
Fee minor comments, but looks great.

File: sdk/state/state.go [hunk 2]
```diff
@@ -43,8 +42,8 @@ type Channel struct {
 	remoteSigner *keypair.FromAddress
 
 	latestCloseAgreement *CloseAgreement
-	// TODO - need to store this too?
-	// lastUnConfirmedPayment *Payment
+	// TODO - set this, probably use different name
+	lastUnConfirmedPayment *Payment
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/state.go]: https://github.com/stellar/experimental-payment-channels/pull/77/commits/988a4fd807125821fadb7cb4cb2bc6e1bd6c4e3a

File: sdk/state/state.go [hunk 3]
```diff
@@ -75,9 +74,16 @@ func NewChannel(c Config) *Channel {
...
+func (c *Channel) IterationNumber() int64 {
+	var latestI int64
+	if c.lastUnConfirmedPayment != nil {
+		latestI = c.lastUnConfirmedPayment.IterationNumber
+	} else if c.latestCloseAgreement != nil {
+		latestI = c.latestCloseAgreement.IterationNumber
+	} else {
+		latestI = 0
+	}
+	return latestI + 1
```
[the remark this answers is not in the document]
Comment (acharb) [reply sdk/state/state.go:86]: https://github.com/stellar/experimental-payment-channels/pull/77/commits/e0be1bf97357c85e630782119c0720a7dc671ae7

File: sdk/state/update.go (hunk 1 above)
Comment (acharb) [reply sdk/state/update.go]: https://github.com/stellar/experimental-payment-channels/pull/77/commits/9b23f3e2e318b766708891ed4bc0711c2c850493

**Pull request merged**